#1083 only need username for joind in link

chrisforrence opened the pull request:
This pull request addresses and closes #1083 by taking the following actions:

- **Database migration**: Adds a new `joindin_username` column, breaks the old `url` column apart and adds the username portion to the new column, then clears the old `url` column (rollback is to take the `joindin_username`, prepend it with "https://joind.in/user/" and put it back into the `url` column before dropping the column)
- **Form**: Adds a new "joind.in username" field with relevant validation
- **Form**: Re-uses the `url` field to store any URL that the user wishes (e.g. personal sites, LinkedIn)

As a note, I had to use a getter method in the dashboard.twig file in order to retrieve the new field; had I missed something? How would I make it so it would just be accessible by the property alone?

localheinz commented:
@chrisforrence 

Before I review this, following the merge of #1165, can you rebase this on top of `master`?

While at it, what do you think about squashing a few commits here and there?

chrisforrence commented:
Can do; I’ll also fix a potential issue I had made in the migration
rollback.

On Wed, Oct 3, 2018 at 5:43 AM Andreas Möller <notifications@github.com>
wrote:

> @chrisforrence <https://github.com/chrisforrence>
>
> Before I review this, following the merge of #1165
> <https://github.com/opencfp/opencfp/pull/1165>, can you rebase this on
> top of master?
>
> While at it, what do you think about squashing a few commits here and
> there?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/opencfp/opencfp/pull/1166#issuecomment-426574384>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABSPPaDsdiDTEPw-fKNJwcQ7UvJeh4jeks5uhIbYgaJpZM4XFKlB>
> .
>
-- 
Cheers!
Chris Forrence

chrisforrence commented:
Alright @localheinz, I went ahead and rebased/squashed, fixed the rollback issue I alluded to earlier. The build appears to have timed out; is that something I can re-kickoff, or is that something you do?

localheinz commented:
@chrisforrence 

Apologies for the delay with my response, I had a bit of a 🤧 last week, and then I was a bit busy attending ScotlandPHP. Taking another look now!

localheinz commented on migrations/20181002234013_create_user_joind_in_username_column.php:
```diff
@@ -0,0 +1,73 @@
+        // Go through each record in user, strip out (https://joind.in/user/) and copy to joindin_username
+        $joindin_regex = '/^https:\/\/joind\.in\/user\/(.{1,100})$/';
+
+        $users = EloquentUser::all();
```
Personally, I'm not a fan of using an ORM in database migrations, as when the migration is run

* a model or entity may not even exist anymore
* a model or entity might not even work with the schema at the state before the migration is run

Now, since I haven't used Laravel at all, and apparently models have been used in migrations within this project before, so I'm not going to make an issue out of it, but for the record, I would prefer not to use models or entities in a database migration.

* at the time the migration is run,

chartjes replied:
We are already using Eloquent in other migrations, so I do not have a big problem with it.

localheinz commented on migrations/20181002234013_create_user_joind_in_username_column.php:
```diff
@@ -0,0 +1,73 @@
+        foreach ($users as $user) {
+            $user->url = $user->joindin_username
+                                        ? 'https://joind.in/user/' . $user->joindin_username
+                                        : '';
```
The `url` column is currently nullable, should this be `null` instead of `''`?

chartjes replied:
Yes, should be `null`

localheinz commented on migrations/20181002234013_create_user_joind_in_username_column.php:
```diff
@@ -0,0 +1,73 @@
+            $user->url = $user->joindin_username
+                                        ? 'https://joind.in/user/' . $user->joindin_username
+                                        : '';
+            $user->joindin_username = '';
```
Should this be `null` instead?

Doesn't make a lot of difference since we drop the column below anyway.

chartjes replied:
Should be `null` for consistency's sake

chrisforrence commented:
That's quite alright @localheinz; hope you're feeling better!

localheinz commented on migrations/20181002234013_create_user_joind_in_username_column.php:
```diff
@@ -0,0 +1,73 @@
+            ->update();
+
+        // Go through each record in user, strip out (https://joind.in/user/) and copy to joindin_username
+        $joindin_regex = '/^https:\/\/joind\.in\/user\/(.{1,100})$/';
```
What do you think about using `camelCase` here instead of `snake_case`?

chartjes replied:
Yes, `camelCase` please

localheinz commented on resources/views/admin/speaker/view.twig:
```diff
@@ -15,6 +15,15 @@
+                {% if speaker.getJoindInUsername() %}
+                    <span class="mr-3">
+                        <a class="hover:text-brand" target="_blank"
+                           href="https://joind.in/user/{{ speaker.getJoindInUsername() }}"
```
Unless I am terribly mistaken, and to answer your question from the body of the pull request, I think it should be possible to use 

```diff
diff --git a/resources/views/admin/speaker/view.twig b/resources/views/admin/speaker/view.twig
index ace150ce..3d9e23fe 100644
--- a/resources/views/admin/speaker/view.twig
+++ b/resources/views/admin/speaker/view.twig
@@ -18,9 +18,9 @@
                 {% if speaker.getJoindInUsername() %}
                     <span class="mr-3">
                         <a class="hover:text-brand" target="_blank"
-                           href="https://joind.in/user/{{ speaker.getJoindInUsername() }}"
+                           href="https://joind.in/user/{{ speaker.joindInUsername }}"
                            rel="noopener noreferrer">
-                            <i class="fa fa-commenting-o"></i> {{ speaker.getJoindInUsername() }}
+                            <i class="fa fa-commenting-o"></i> {{ speaker.joindInUsername }}
                         </a>
                     </span>
                 {% endif %}
```

here instead. Have you tried?

Maybe it also makes sense to extract a method and shift the construction of the URL to `SpeakerProfile`, what do you think?

chartjes replied:
I like the idea of having `SpeakerProfile` generate this too since I recently merged in a PR that moved the generation of Twitter URL's to `SpeakerProfile` as well

chrisforrence replied:
Haha, I spent half an hour on stream trying to figure out why `speaker.joindInUsername` hadn't been working :p I can certainly give it another crack tonight though!

I can definitely be on board with either constructing the URL through SpeakerProfile or on a JoindIn helper (and something similar for Twitter as well)

chrisforrence replied:
Ah, well then, that answers the second part!

localheinz commented on src/Http/Form/SignupForm.php:
```diff
@@ -201,15 +202,29 @@ public function validateLastName(): bool
     public function validateUrl(): bool
     {
-        if (\preg_match('/^https:\/\/joind\.in\/user\/[a-zA-Z0-9\-_\.]{1,100}$/', $this->cleanData['url'])
+        if (\filter_var($this->cleanData['url'], FILTER_VALIDATE_URL) !== false
```
👍

localheinz commented on src/Http/Form/SignupForm.php:
```diff
@@ -201,15 +202,29 @@ public function validateLastName(): bool
+    {
+        if (!isset($this->cleanData['joindin_username'])
+            || $this->cleanData['joindin_username'] == ''
+            || \preg_match('/^[a-zA-Z0-9\-_\.]{1,100}$/', $this->cleanData['joindin_username'])
```
Should we make sure we are dealing with a `string` here before applying `preg_match()`?

localheinz commented on tests/Unit/Domain/Speaker/SpeakerProfileTest.php:
```diff
@@ -273,6 +273,40 @@ public function getTwitterReturnsTwitterIfPropertyIsNotHidden()
+     */
+    public function getJoindInUsernameReturnsJoindInUsernameIfPropertyIsNotHidden()
+    {
+        $joindin_username = $this->faker()->userName;
```
Can we use `camelCase` here instead of `snake_case`?

chartjes replied:
Yep, `camelCase` please

localheinz commented on tests/Unit/Http/Form/SignupFormTest.php:
```diff
@@ -286,7 +286,58 @@ public function lastNameProvider(): array
 
     /**
-     * Test that the Joind.in URL is being validated correctly
+     * Test that the joind.in username is being validated correctly
```
Not sure, perhaps we can even remove the short description?

chartjes replied:
Agreed

localheinz commented on tests/Unit/Http/Form/SignupFormTest.php:
```diff
@@ -286,7 +286,58 @@ public function lastNameProvider(): array
+     * Data provider for joindInUsernameIsValidatedCorrectly
+     *
+     * @return array
+     */
```
Perhaps we can remove the entire DocBlock?

chartjes replied:
Yeah, not required for tests as far as I am concerned

localheinz commented on tests/Unit/Http/Form/SignupFormTest.php:
```diff
@@ -286,7 +286,58 @@ public function lastNameProvider(): array
+        $longUsername = '';
+
+        for ($x = 1; $x <= 256; ++$x) {
+            $longUsername .= 'X';
```
I think this can easily be replace with

```php
$longUserName = str_pad('', 255, 'X');
```

or

```
$longUsername = str_repeat('X', 255);
```

Ha, sorry, I just realized it's moved from existing code. Do you mind adjusting, though?

Maybe it would even be more interesting to test the edges?

* length `100` is ok
* length `101` is not ok

What do you think?

chrisforrence replied:
Yeah, I can make adjustments here tonight!

localheinz commented on migrations/20181002234013_create_user_joind_in_username_column.php:
```diff
@@ -0,0 +1,73 @@
+    /** @var Capsule $capsule */
+    public $capsule;
+
+    public function bootEloquent()
```
What do you think about adding return type declarations (as far as we can) here and at every newly added method?

chartjes replied:
Yes, go all in on return type declarations

chrisforrence commented:
I messed up rebasing a bit; I ended up cutting a new branch, taking the diff of upstream/master..HEAD and creating a new branch. I'll submit that one as a pull request; I'm sorry about that.
